Code/src/main/java/nl/utwente/viskell/ui/components/Block.java

```diff
@@ -151,46 +162,73 @@ public void handleConnectionChanges(boolean finalPhase) {
-        LetExpression fullExpr = new LetExpression(this.getLocalExpr());
-        this.extendExprGraph(fullExpr);
+        Pair<Expression, Set<Block>> localExpr = getLocalExpr();
+        
+        LetExpression fullExpr = new LetExpression(localExpr.a, false);
+        Set<Block> surroundingBlocks = localExpr.b;
+        extendExprGraph(fullExpr, null, surroundingBlocks);
+        
+        surroundingBlocks.forEach(block -> block.extendExprGraph(fullExpr, null, new HashSet<>()));
+        
         return fullExpr;
     }
 
     /**
      * Extends the expression graph to include all subexpression required
      * @param exprGraph the let expression representing the current expression graph
+     * @param container the container to which this expression graph is constrained
+     * @param addLater a mutable list of blocks that have to be added by a surrounding container
      */
-    protected void extendExprGraph(LetExpression exprGraph) {
+    protected void extendExprGraph(LetExpression exprGraph, BlockContainer container, Set<Block> addLater) {
          for (InputAnchor input : this.getAllInputs()) {
-             input.extendExprGraph(exprGraph);
+             input.extendExprGraph(exprGraph, container, addLater);
          }
     }
     
-    /**
-     * Called when the VisualState changed.
-     */
+    /** Called when the VisualState changed. */
     public abstract void invalidateVisualState();
 
-    /** 
-     * @return whether this block is visually shown below common blocks (is constant per instance).
-     */
+    /** @return whether this block is visually shown below common blocks (is constant per instance). */
     public boolean belongsOnBottom() {
         return false;
     }
     
-    /**
-     * @return class-specific properties of this Block.
-     */
+    /** @return class-specific properties of this Block. */
     protected ImmutableMap<String, Object> toBundleFragment() {
         return ImmutableMap.of();
     }
+    
+    /** @return the container to which this block belongs, if any */
+    public Optional<BlockContainer> getContainer() {
+        return container;
+    }
+    
+    /** Removes the block from its container */
+    public void detachFromContainer() {
+        container.ifPresent(container -> {
+            this.container = Optional.empty();
+            container.detachBlock(this);
+        });
+    }
 
+    /** Scans for and attaches to a new container, if any */
+    public void refreshContainer() {
+        Bounds myBounds = localToScene(getBoundsInLocal());
+        detachFromContainer();
+        
+        container = parentPane.getBlockContainers().
+            filter(container -> container.localToScene(container.getBoundsInLocal()).contains(myBounds)).
+                reduce((a, b) -> !a.localToScene(a.getBoundsInLocal()).contains(b.localToScene(b.getBoundsInLocal())) ? a : b);
+        
+        container.ifPresent(container -> container.attachBlock(this));
```

Comment:
Minor nitpick: naming the lambda-local variable the same as the field is a bit confusing.